File: examples/typescript/agent-checkout/src/checkout.ts
// copyright © 2025-2026 bkey inc. all rights reserved.

/**
 * TypeScript agent-checkout example.
 *
 * Shows both patterns side-by-side:
 *
 *   1. Generic one-line CIBA approval — `bkey.approve(...)` — the universal
 *      primitive for any sensitive action (deploy, refund, DB drop, admin
 *      grant, etc.). Returns an EdDSA-signed JWT proving user consent.
 *
 *   2. Structured checkout — `bkey.createCheckoutRequest(...)` +
 *      `pollCheckoutRequest(...)` — the same CIBA primitive with checkout-
 *      specific fields (merchant, items, amount) rendered on the phone.
 *
 * Run:
 *   npm run dev
 */

import 'dotenv/config';
import { BKey, pollCheckoutRequest } from '@bkey/sdk';

const BKEY_API_URL = process.env.BKEY_API_URL ?? 'https://api.bkey.id';
const BKEY_CLIENT_ID = process.env.BKEY_CLIENT_ID;
const BKEY_CLIENT_SECRET = process.env.BKEY_CLIENT_SECRET;
const BKEY_USER_DID = process.env.BKEY_USER_DID;

if (!BKEY_CLIENT_ID || !BKEY_CLIENT_SECRET || !BKEY_USER_DID) {
  console.error(
    'Missing credentials. Set BKEY_CLIENT_ID, BKEY_CLIENT_SECRET, ' +
      'BKEY_USER_DID in .env (copy from .env.example).',
  );
  process.exit(1);
}

const bkey = new BKey({
  apiUrl: BKEY_API_URL,
  clientId: BKEY_CLIENT_ID,
  clientSecret: BKEY_CLIENT_SECRET,
  did: BKEY_USER_DID,
});

async function demoGenericApproval(): Promise<void> {
  console.log('='.repeat(60));
  console.log('Part 1: Generic biometric approval');
  console.log('='.repeat(60));
  console.log('Check your phone — Face ID prompt incoming.\n');

  const result = await bkey.approve('Proceed with a test action', {
    scope: 'approve:action',
  });

  if (!result.approved) {
    console.log('Denied on device. Aborting.');
    process.exit(1);
  }

  console.log(`Approved. JWT prefix: ${result.accessToken.slice(0, 24)}...`);
  console.log('Verify this token server-side via /oauth/jwks before acting.\n');
}

async function demoCheckout(): Promise<void> {
  console.log('='.repeat(60));
  console.log('Part 2: Structured checkout');
  console.log('='.repeat(60));

  const created = (await bkey.createCheckoutRequest({
    merchantName: 'BKey Demo Store',
    merchantDomain: 'demo.bkey.id',
    checkoutUrl: 'https://demo.bkey.id/checkout/example',
    amount: 39.97,
    currency: 'USD',
    lineItems: [
      { title: 'Widget', quantity: 1, price: 9.99 },
      { title: 'Gadget', quantity: 2, price: 14.99 },
    ],
  })) as { checkoutRequest: { id: string } };

  const checkoutId = created.checkoutRequest.id;
  console.log(`Checkout created: ${checkoutId}`);
  console.log('Phone shows a shopping-cart-shaped approval screen.\n');

  const outcome = await pollCheckoutRequest(bkey, checkoutId, 180_000);
  console.log(`Checkout ${outcome.status}.`);
  if (outcome.orderConfirmation) {
    console.log('Order confirmation:', JSON.stringify(outcome.orderConfirmation, null, 2));
  }
}

async function main(): Promise<void> {
  try {
    await demoGenericApproval();
    await demoCheckout();
  } catch (err) {
    console.error('Example failed:', err instanceof Error ? err.message : err);
    process.exit(1);
  }
}

void main();

File: examples/python/agent-checkout/checkout.py
"""Python agent-checkout example.

Demonstrates both patterns side-by-side:

  1. Generic one-line CIBA approval — the universal pattern for any
     sensitive action (deploy, refund, DB write, admin grant, etc.).

  2. Structured checkout — the same CIBA primitive with checkout-specific
     fields (merchant, items, amount) rendered on the user's phone.

Prereqs:
  - Install the BKey CLI: npm install -g @bkey/cli
  - Create agent credentials: bkey auth login && bkey auth setup-agent --save
  - Copy .env.example → .env and fill in the values.

Run:
  python checkout.py
"""

import os
import sys
from pathlib import Path

from bkey import BKeyClient
from bkey.exceptions import ApprovalDeniedError, ApprovalTimeoutError


def _load_env() -> None:
    env_path = Path(__file__).with_name(".env")
    if not env_path.exists():
        return
    for line in env_path.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        os.environ.setdefault(key.strip(), value.strip().strip('"').strip("'"))


def main() -> None:
    _load_env()

    client_id = os.environ.get("BKEY_CLIENT_ID")
    client_secret = os.environ.get("BKEY_CLIENT_SECRET")
    user_did = os.environ.get("BKEY_USER_DID")
    base_url = os.environ.get("BKEY_API_URL", "https://api.bkey.id")

    if not (client_id and client_secret and user_did):
        print(
            "Missing credentials. Set BKEY_CLIENT_ID, BKEY_CLIENT_SECRET, "
            "BKEY_USER_DID in .env (copy from .env.example)."
        )
        sys.exit(1)

    client = BKeyClient(
        client_id=client_id,
        client_secret=client_secret,
        base_url=base_url,
    )

    # ── Part 1: Generic CIBA approval (one line) ──────────────────────
    print("=" * 60)
    print("Part 1: Generic biometric approval")
    print("=" * 60)
    print("Requesting approval on the user's phone (check Face ID prompt)...")
    try:
        result = client.approve(
            message="Proceed with a test action (example agent)",
            user_did=user_did,
            scope="approve:action",
        )
    except ApprovalDeniedError:
        print("Denied on device. Exiting.")
        return
    except ApprovalTimeoutError:
        print("No response within timeout. Exiting.")
        return

    print(f"Approved. Access token starts with: {(result.access_token or '')[:24]}...")
    print("(Verify this JWT server-side with the /oauth/jwks endpoint before acting on it.)\n")

    # ── Part 2: Structured checkout (same CIBA primitive, richer UI) ──
    print("=" * 60)
    print("Part 2: Structured checkout")
    print("=" * 60)
    checkout = client.checkout_request(
        merchant_name="BKey Demo Store",
        items=[
            {"name": "Widget", "price": 9.99, "quantity": 1},
            {"name": "Gadget", "price": 14.99, "quantity": 2},
        ],
        amount=39.97,
        currency="USD",
    )
    print(f"Checkout initiated: {checkout.id}")
    print("Approve on the phone. The screen shows merchant + items + total.\n")

    try:
        outcome = client.checkout_poll(checkout.id)
    except ApprovalDeniedError:
        print("Customer rejected the checkout.")
        return
    except ApprovalTimeoutError:
        print("Checkout expired without a decision.")
        return

    print(f"Checkout {outcome.status}.")
    if outcome.payment_intent_id:
        print(f"Payment intent: {outcome.payment_intent_id}")


if __name__ == "__main__":
    main()
